Apply Disable ExtAuth from GlobalExtAuth and Remove Auth from HTTP Upgrade

SamMHD opened the pull request:
closes #6617 and #6659

--- 

In order to fix #6617 I have changed how `dagRoute.DisableAuth` is caclulated. 

Also, to fix #6659 I thought it is needed to change behavior when we are upgrading a request to HTTPS (when `permitInsecure` is not set and we are redirecting HTTP to HTTPS). Thereby, when `dagRoute.AuthContext`/`dagRoute.AuthDisabled` is configured it will affect redirection as well --  before these changes, redirection won't care about them.

---

Changes:
- use dagRoute's AuthContext and AuthDisabled in HTTPS-Upgrade to fix 6659
- Use globalExtAuth.AuthPolicy.Disabled to calculate dagRoute.AuthDisabled
- Fix Tests

SamMHD commented:
@sunjayBhatia can you please add required labels to this one?
cc: @skriss @rajatvig @davinci26

SamMHD commented:
Thank you @tsaarni for the label.
When can we proceed to merge?

tsaarni commented:
@SamMHD I will try to arrange time for myself to review ASAP today or tomorrow!

SamMHD commented:
Thank you so much @tsaarni

tsaarni reviewed (changes requested):
Looks good!

Regarding testing the fix: Would it be possible to add a new test case in the e2e test suite that would have caught this bug. There are global ext auth tests [here](https://github.com/projectcontour/contour/blob/39544fec44712b78cb45c9585cb903323420c446/test/e2e/httpproxy/httpproxy_test.go#L813-L859) but they currently only cover `disabled: false`.

Also tagging @clayton-gonsalves for your input, as your review would be valuable since you originally implemented this functionality.

SamMHD commented on internal/envoy/v3/route.go:
```diff
@@ -121,6 +121,15 @@ func buildRoute(dagRoute *dag.Route, vhostName string, secure bool) *envoy_confi
 		// envoy.RouteRoute. Currently the DAG processor adds any HTTP->HTTPS
 		// redirect routes to *both* the insecure and secure vhosts.
 		route.Action = UpgradeHTTPS()
+
+		route.TypedPerFilterConfig = map[string]*anypb.Any{}
+		// Apply per-route authorization policy modifications.
+		if dagRoute.AuthDisabled {
+			route.TypedPerFilterConfig["envoy.filters.http.ext_authz"] = routeAuthzDisabled()
+		} else if len(dagRoute.AuthContext) > 0 {
+			route.TypedPerFilterConfig["envoy.filters.http.ext_authz"] = routeAuthzContext(dagRoute.AuthContext)
+		}
+
 	case dagRoute.DirectResponse != nil:
```
Yes, I agree with you. we can even disable it totally. But right now it's configurable per route and one can enable/disable it as per their need.
But if you think that it will increase complexity of Contour's behavior I can disable it at all.

tsaarni replied:
I did not test this so I'm not 100% sure, but I'm assuming that currently two calls to authorization service will be generated for single client when client goes through the HTTPS upgrade? Once when hitting that route on the HTTP listener, and second time after redirect when the client accesses the actual route on HTTPS listener. If that happens, it sounds bit surprising to me. I would have thought there should be just single call, since the backing resource is accessed only once - and that was why I was thinking that should the HTTPS redirect be done without authorization. What do you think?

SamMHD replied:
I agree with you and I think we can disable these calls (before redirections).

I'll change the code.

SamMHD replied:
@tsaarni, It is fixed 🤝
You can review.

SamMHD commented:
Thank you for your feedback; I'll add the tests as suggested. 

I previously attempted to write tests for this scenario but found the existing test functions a bit incompatible to implement `disabled: true` for GlobalAuth, which led me to initially omit them. Now that you've requested it, I'll revisit this and include the necessary tests.

tsaarni commented:
> I previously attempted to write tests for this scenario but found the existing test functions a bit incompatible to implement `disabled: true` for GlobalAuth, which led me to initially omit them. Now that you've requested it, I'll revisit this and include the necessary tests.

Exactly, the existing test functions make no sense for `disabled: true`. There could be a new one to test this behaviour specifically.

SamMHD commented:
> Exactly, the existing test functions make no sense for `disabled: true`. There could be a new one to test this behaviour specifically.

I think I have implemented a workaround to this by making the test function more flexible. 
I'll push the changes now, hope you find time to review this.

SamMHD commented:
@tsaarni Do have time to review the changes again?

SamMHD commented:
@tsaarni @sunjayBhatia @skriss is there any update on this?

SamMHD commented:
Why nobody answer us? :((( 
@tsaarni @sunjayBhatia @skriss

SamMHD commented:
thank you @tsaarni , when do you think you can give it a look?

SamMHD commented:
Hello again, Is there any update on this PR? @tsaarni

SamMHD commented:
Very good point dear @tsaarni , I have visited that code branch and fixed it, other than this one more thing came to my attention, and it was that "If we disable global auth in vhost configuration then we can not enable it back again in a route" so I have fixed it in "else clause" of same code branch.

Can you give it a try or help me setup your test environment for myself? I think it should be fixed now.

SamMHD commented:
dear @tsaarni , do you have any update on this PR?

izturn commented:
/lgtm

SamMHD commented:
Thank you so much @izturn  🙌
Are we going to merge or we need additional approvals?

tsaarni commented on changelogs/unreleased/6661-SamMHD-minor.md:
```diff
@@ -0,0 +1,7 @@
+## Disable ExtAuth by default if GlobalExtAuth.AuthPolicy.Disabled is set
+
+Global external authorization or vhost-level authorization is enabled by default unless an AuthPolicy explicitly disables it. By default, `disabled` is set to `GlobalExtAuth.AuthPolicy.Disabled`. This global setting can be overridden by vhost-level AuthPolicy, which can further be overridden by route-specific AuthPolicy. Therefore, the final authorization state is determined by the most specific  ...
```
```suggestion
Global external authorization can now be disabled by default and enabled by overriding the vhost and route level auth policies.
This is achieved by setting the `globalExtAuth.authPolicy.disabled` in the configuration file or `ContourConfiguration` CRD to `true`, and setting the `authPolicy.disabled` to `false` in the vhost and route level auth policies.
The final authorization state is determined by the most specific policy applied at the route level.
```

tsaarni commented on changelogs/unreleased/6661-SamMHD-minor.md:
```diff
@@ -0,0 +1,7 @@
+
+Global external authorization or vhost-level authorization is enabled by default unless an AuthPolicy explicitly disables it. By default, `disabled` is set to `GlobalExtAuth.AuthPolicy.Disabled`. This global setting can be overridden by vhost-level AuthPolicy, which can further be overridden by route-specific AuthPolicy. Therefore, the final authorization state is determined by the most specific  ...
+
+## Disable External Authorization in UpgradeHTTPS
+
+From now on, Contour will configure Envoy to handle HTTPS Redirection without authorization on routes. (previously if GlobalExtAuth was set, Envoy would check request with ext_auth before redirection which could result in 401 instead of redirection)
```
```suggestion
## Disable External Authorization in HTTPS Upgrade

When external authorization is enabled, no authorization check will be performed for HTTP to HTTPS redirection.
Previously, external authorization was checked before redirection, which could result in a 401 Unauthorized error instead of a 301 Moved Permanently status code.
```

tsaarni reviewed (commented):
@SamMHD Looks very good! I re-tested the PR and it worked perfectly now.  I still added suggestions for the change log to hopefully highlight the change bit more clearly to the users, please check and see what you think.

SamMHD commented:
Thank you dear @tsaarni for the changes 🙏 Applied all of your suggestions ✅

SamMHD commented:
Will we proceed to merge or we need more approvals?

sunjayBhatia commented:
> Will we proceed to merge or we need more approvals?

Could you add a more descriptive PR title please? thanks!

SamMHD commented:
> > Will we proceed to merge or we need more approvals?
> 
> Could you add a more descriptive PR title please? thanks!

@sunjayBhatia , I have changed PR title. Is it good to go or you recommend another title?

SamMHD commented:
btw, Sorry @tsaarni for that force push, I had to sign-off to pass DCO

tsaarni commented:
> btw, Sorry @tsaarni for that force push, I had to sign-off to pass DCO

No problem @SamMHD, github luckily shows the commit hash before and after force push so it becomes easy to spot the changes :)

tsaarni reviewed (approved):
Thank you @SamMHD for your contribution and your patience with all the review delays!

SamMHD commented:
Thank you @tsaarni @sunjayBhatia @izturn for the efforts and review 🙏
